app/src/main/java/org/dhis2/usescases/searchTrackEntity/SearchTEIViewModel.kt

```diff
@@ -0,0 +1,268 @@
+        )
+    }
+
+    fun setSearchScreen() {
+        _screenState.value = SearchForm(
+            previousSate = _screenState.value?.screenState ?: SearchScreenState.NONE,
+            queryHasData = queryData.isNotEmpty(),
+            minAttributesToSearch = _selectedProgram.value?.minAttributesRequiredToSearch() ?: 0
+        )
+    }
+
+    fun setPreviousScreen() {
+        when (_screenState.value?.previousSate) {
+            SearchScreenState.LIST -> setListScreen()
+            SearchScreenState.MAP -> setMapScreen()
+            SearchScreenState.SEARCHING -> setSearchScreen()
+            SearchScreenState.ANALYTICS -> setAnalyticsScreen()
+            else -> {
+            }
+        }
+    }
+
+    fun refreshData() {
+        onSearchClick()
+    }
+
+    fun updateQueryData(rowAction: RowAction) {
+        if (rowAction.value != null) {
+            queryData[rowAction.id] = rowAction.value!!
+        } else {
+            queryData.remove(rowAction.id)
+        }
+        updateSearch()
+    }
+
+    fun clearQueryData() {
+        queryData.clear()
+        updateSearch()
+    }
+
+    fun restoreQueryData(queryDataBackUp: HashMap<String, String>) {
+        queryData.clear()
+        queryData.putAll(queryDataBackUp)
+        updateSearch()
+    }
+
+    private fun updateSearch() {
+        if (_screenState.value is SearchForm) {
+            _screenState.value =
+                (_screenState.value as SearchForm).copy(queryHasData = queryData.isNotEmpty())
+        }
+    }
+
+    fun fetchListResults(): LiveData<PagedList<SearchTeiModel>>? {
+        return when {
+            searching -> {
+                searchRepository.searchTrackedEntities(
+                    SearchParametersModel(
+                        selectedProgram = _selectedProgram.value,
+                        queryData = queryData
+                    ),
+                    searching && networkUtils.isOnline()
+                )
+            }
+            displayFrontPageList() -> {
+                searchRepository.searchTrackedEntities(
+                    SearchParametersModel(
+                        selectedProgram = _selectedProgram.value,
+                        queryData = queryData
+                    ),
+                    false
+                )
+            }
+            else -> {
+                null
+            }
+        }
+    }
+
+    fun fetchGlobalResults(): LiveData<PagedList<SearchTeiModel>>? {
+        return if (searching) {
+            searchRepository.searchTrackedEntities(
+                SearchParametersModel(
+                    selectedProgram = null,
+                    queryData = queryData
+                ),
+                searching && networkUtils.isOnline()
+            )
+        } else {
+            null
+        }
+    }
+
+    fun fetchMapResults() {
+        viewModelScope.launch {
+            val result = async(context = Dispatchers.IO) {
```

Comment:
You will need to Inject Dispatchers IO for testing it, and passing an Undefined.

Reply:
Done! Thanks for the reminder! :)